analysis/topEFT/dataDrivenEstimation.py

```diff
@@ -0,0 +1,106 @@
+import argparse
+from coffea import hist, processor
+from topcoffea.modules.YieldTools import YieldTools
+from topcoffea.modules.GetValuesFromJsons import get_lumi
+import cloudpickle
+
+class DataDrivenProducer: 
+    def __init__(self, inputHist, outputName, year, doDDFakes=True, doDDFlips=False):
+        self.yt=YieldTools()
+        if type(inputHist) == str: # we are plugging a pickle file
+            self.inhist=yt.get_hist_from_pkl(inputHist)
+        else: # we already have the histogram
+            self.inhist=inputHist
+        self.outputName=outputName
+        if doDDFlips: 
+            raise RuntimeError("Data driven flips not yet implemented")
+        self.doDDFlips=doDDFlips
+        self.doDDFakes=doDDFakes
+        self.dataName='data'
+        self.outHist=None
+        self.year=year
+
+        self.DDFakes()
+
+    def DDFakes(self):
+        if self.outHist!=None:  # already some processing has been done, so using what is available
+            self.inhist=self.outHist
+            
+        self.outHist={}
```

Comment:
Have you tested this case? You might need to use `self.inhist=self.outHist.copy()` because Python usually uses references, so setting `self.outHist={}` _might_ also clear `self.inhist`.

Reply:
I haven't tested this case but an equivalent MWE:
>>> a={'a':'a'}
>>> b=a
>>> a={}
>>> b
{'a': 'a'}

so I think it shouldn't be a problem

Reply:
Ok. Initially I was afraid about the quadratic fits, but since these are only backgrounds, we don't use `HistEFT`.

Reply:
You are correct that python stores references to non-primitive types (basically anything more complicated than a string/int/float), however, you're confusing variable assignment with object modification.

The above code shouldn't result in clearing `self.inhist` b/c what's actually going on is that `self.inhist=self.outHist` is _assigning_ the variable named `self.inhist` to the _object_ that is referred to by `self.outHist`, so now both variables refer to the same object in memory. Now when we later set `self.outHist = {}` that is assigning this variable to an entirely new object (a new instantiated and empty dictionary), however, `self.inhist` still points to the old object in memory.

The case that you're thinking of that you need to be careful is when you assign an object to two different variables and then you _modify_ that object using one of those variables as this will change the actual object in memory and so both variables (which are pointing to the same object) will reflect this change. So for example if `self.outHist={}` was instead `self.outHist.clear()`, that _would_ erase the object that is referred to by `self.inhist` since we're modifying the actual object in memory, rather than just point to a completely new object.